fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,470 @@
+    else:
+        remaining_ffg_weight = 0
+
+    max_adversarial_ffg_support_for_checkpoint = int(
+        min(
+            ceil_div(total_active_balance * CONFIRMATION_BYZANTINE_THRESHOLD, 100),
+            CONFIRMATION_SLASHING_THRESHOLD,
+            ffg_support_for_checkpoint
+        )
+    )
+
+    # Returns whether the following condition is true using only integer arithmetic
+    # 2 / 3 * total_active_balance <= (
+    #     ffg_support_for_checkpoint - max_adversarial_ffg_support_for_checkpoint +
+    #     (1 - CONFIRMATION_BYZANTINE_THRESHOLD / 100) * remaining_ffg_weight
+    # )
+
+    return (
+        200 * total_active_balance <=
+        ffg_support_for_checkpoint * 300 + (300 - 3 * CONFIRMATION_BYZANTINE_THRESHOLD) *
+        remaining_ffg_weight - max_adversarial_ffg_support_for_checkpoint * 300
+    )
+```
+
+### `is_confirmed_no_caching`
+
+```python
+def is_confirmed_no_caching(store: Store, block_root: Root) -> bool:
+    current_epoch = get_current_store_epoch(store)
+
+    block = store.blocks[block_root]
+    block_state = store.block_states[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+
+    if current_epoch == block_epoch:
+        return (
+            get_checkpoint_block(store, block_root, Epoch(current_epoch - 1)) == 
+            block_state.current_justified_checkpoint.root
+            and is_lmd_confirmed(store, block_root)
+            and is_ffg_confirmed(store, block_root, current_epoch)
+        )
+    else:
+        return (
+            compute_slots_since_epoch_start(get_current_slot(store)) == 0
+            and is_ffg_confirmed(store, block_root, Epoch(current_epoch - 1))
+            and any(
+                [
+                    get_voting_source(store, leaf).epoch + 2 >= current_epoch
+                    and is_lmd_confirmed(store, leaf)
+                    for leaf in store.leaves_last_slot_previous_epoch
+                ]
+            )
+        )
+```
+
+### `is_confirmed`
+
+```python
+def is_confirmed(store: Store, block_root: Root) -> bool:
+    highest_confirmed_root_since_last_epoch = (
+        store.highest_confirmed_block_current_epoch
+        if store.blocks[store.highest_confirmed_block_current_epoch].slot >
+        store.blocks[store.highest_confirmed_block_previous_epoch].slot
+        else store.highest_confirmed_block_previous_epoch)
+
+    return get_ancestor(store, highest_confirmed_root_since_last_epoch, store.blocks[block_root].slot) == block_root
+```
+
+#### `find_confirmed_block`
+
+```python
+def find_confirmed_block(store: Store, block_root: Root) -> Root:
+
+    if block_root == store.finalized_checkpoint.root:
+        return block_root
+
+    block = store.blocks[block_root]
+    current_epoch = get_current_store_epoch(store)
+
+    block_epoch = compute_epoch_at_slot(block.slot)
+
+    # If `block_epoch` is not either the current or previous epoch, then return `store.finalized_checkpoint.root`
```

Comment:
What is the confirmation rule for the block between finalized checkpoint and justified checkpoint?

Reply:
Any block descendant of the latest finalized checkpoint is treated in the same way